File: PythonAlgorithmInterView/그래프/수열.js
/**
 * @param {number[]} nums
 * @return {number[][]}
 * 46번 문제.
 */
 var permute = function(nums = []) {

    let result = []
    let prev_ele = [];

    let dfs = (elements) => {

        if(elements.length == 0){
            result.push(prev_ele.slice());
            return
        }


        for( let i =0; i < elements.length; i++){
            let next_ele = elements.slice();
            next_ele.splice(i,1);
            
            prev_ele.push(elements[i]);
            dfs(next_ele);
            prev_ele.pop()
        }
    }

    dfs(nums)

    return result
};


let nums = [1,2,3];

console.log(permute(nums))

File: PythonAlgorithmInterView/그래프/수열.py
def permute(self, nums: list[int]) -> list[list[int]]:
    result = []
    prev_list = []

    def dfs(elements):
        if len(elements) == 0:
            result.append(prev_list[:])

        for e in elements :
            next_list = elements[:]
            next_list.remove(e)

            prev_list.append(e)
            dfs(next_list)
            prev_list.pop()


    dfs(nums)

    return result

if __name__ == '__main__':
    nums = [1, 2, 3]
    print( permute(None, nums))